Canvas v0.10 and federation readiness

raykyri opened the pull request:
Closes #8918. All tests pass, ready for QA and code review

This PR updates hicommonwealth/commonwealth from Canvas 0.8 to Canvas 0.10, and fully implements a federation node backed by a separate Postgres database.

When creating, editing, or deleting comments, threads, or reactions, it attempts to create the corresponding Canvas action and post it to the database. This may fail if a dependency of the current action wasn't posted to Canvas (e.g. commenting on a thread when the thread was not signed on Canvas); in that case the action will proceed without a signature.

## Full Description of Changes

- Upgrades canvas dependencies from v0.8 -> v0.10
  - Removes unused canvas-node package
  - Replaces CAIP-2 `address` with `did` around the codebase. The DIDs are unchanged from CAIP-2 identifiers except with `did:pkh:` prefixed before them; [did:pkh](https://github.com/w3c-ccg/did-pkh/blob/main/did-pkh-method-draft.md) is just a wrapper around CAIP-2 but this gives us a better standardized way to refer to user accounts as DIDs.
  - Renames `canvas_hash` to `canvas_msg_id`
  - Ensures `canvas_msg_id` is used when comments depend on threads, reactions depend on comments, etc.
- Removes the ENFORCE_SESSION_KEYS feature flag.
  - A Canvas session is now always required when logging in
  - Interacting on the site now always requires a valid session for the currently active address.
  - The client will attempt to sign actions. This will fail (as mentioned above) if dependent data is missing, in which case it will submit the thread/comment/reaction/etc. to the backend without a signature, which will be accepted. But if a signature is provided, it will be verified by the backend, and an error will be thrown if the Action or Session is invalid.
- Minor upgrades
  - Updates @types/react from ^18.0.26 to ^18.3.3
  - Updates prettier to ^3.3.3, to support some syntax used in the update Canvas contract.ts.
  - Updates typescript to use `es2022` module format.
  - Adds verified checkmark to threads, since they were just on comments before.

Note this PR requires a local database called `federation`, or requires `process.env.FEDERATION_POSTGRES_DB_URL` to be set.  To QA, you can create a local federation database the same way you initialized Postgres while setting up Commonwealth:

```
$ createdb federation
```

Or in psql:
```
CREATE DATABASE federation;
```

## Deployment Plan

* Create a new Heroku database and connect it to the application (it could be the larger Hobby database).
* Provide it as an environment variable: FEDERATION_POSTGRES_DB_URL

## Description of Changes

- [x]  Upgraded to 0.10.0
- [x]  Change everything to use DIDs
- [x]  Use ES2022 to fix import
- [x]  Actions inserted when creating a comment, thread, and both types of reactions
    - [x]  `thread`
    - [x]  `comment`
    - [x]  `reactThread`
    - [x]  `reactComment`
- [x]  app.messageLog.verifySignature and app.messageLog.insert not consistent
- [x]  async issue with app.insert()
- [x]  Test on different chains
    - [x]  Ethereum
    - [x]  Evmos (Cosmos Metamask)
    - [x]  Evmos (Keplr)
    - [x]  Cosmos
    - [x]  Solana
    - [x]  Substrate
- [x] Fix references to canvas_hash, by using `msgid` in all canvas actions that reference a different model. Also ensure that server-side verification is implemented and working for all actions.
    - Threads
    - [x] `createThread`
    - [x] `updateThread` (limited server-side verification)
    - [x] `deleteThread`
    - Comments
    - [x] `createComment`
    - [x] `updateComment` (limited server-side verification)
    - [x] `deleteComment`
    - Reactions
    - [x] `reactThread`
    - [x] `reactComment`
    - [x] `unreactComment`
    - [x] `unreactThread`
- [x]  missing fields
    - [x]  `thread.topic`
- [x]  client uses the most recent **clock number + parents** when creating actions/sessions
    - [x]  `/api/getClock` OR just include in app.status
- [x] remove ENFORCE_SESSION_KEYS
- [x] update session revalidation dialog design
- [x] ~~auto-log-out user if all their session keys are logged out~~ (Difficult to do this without either 1) introducing extra latency on page load (bad), 2) refactoring initAppState, or 3) making the code more complicated if the user's session keys are all logged out)
- [x] fix issue where user gets auto-logged when relinking a wallet with a valid cached session key

Todos before deploying:

- [ ] Discuss how to provision and maintain the federation database on QA, any staging tiers, and production tiers.

Outside the scope of this PR:

- [x] Turn libp2p on
- [x] Set up explorer - Create a Commonwealth version of the application at https://github.com/canvasxyz/canvas/pull/290

raykyri commented:
Need to fix tests but otherwise this is ready for review/QA.

jnaviask commented:
@raykyri what does this PR do? where's the description of the *goals* of these changes (should be a ticket, ideally), and how we can confirm that these goals were satisfactorily met?

raykyri commented:
Good point.. I was going to work on this on Sunday but lost track of time. Let me move this into a draft PR until I can fix that up (hopefully later today).

raykyri commented:
@rjwebb Can you take a look at the unit tests/integration tests to fix the remaining failing tests? I've updated the seeders to use DIDs, but:

- I think there are some places where they expect signed Canvas messages to be passed that aren't being passed right now
- The action/session clock encoding isn't working properly inside tests
- There are a couple of weird type errors showing up on `pnpm check-types`, which are probably because of a React version bump or incompatibility between conflicting versions of React; this might be caused by some dependency updates that we made. We might have to try to backtrack in history to figure out what caused this, but fwiw it wasn't showing up until I rebased this branch onto master today. (This is lower priority)

You can run a smaller subset of tests from packages/commonwealth using:

```
INIT_TEST_DB=true NODE_ENV=test pnpm exec vitest --config ../../vite.config.ts --fileParallelism=false --coverage run ./test/integration/api/subscriptions.spec.ts "--allowOnly=false"```

kurtisassad commented on common_knowledge/Federation.md:
````diff
@@ -0,0 +1,38 @@
+In another terminal, use the libp2p multiaddr to start a node:
+
+```
+canvas run libs/shared/src/canvas/runtime/contract.ts --bootstrap-list /ip4/127.0.0.1/tcp/8090/ws/p2p/12D3KooWPbkvKDxziibMP3cxrQP7NNnZCDNfj2cT3Ej3TH23wWBS
````
I cant actually get this step to work, so I have downloaded the canvas cli like:

```bash
 npm install -g @canvas-js/cli
```

And then tried running this line but get:
`Unknown arguments: bootstrap-list, bootstrapList`

Trying to change it to --bootstrap, I get:
```
Contract files must match *.canvas.js
```

For reference I tried this with the latest canvas cli as well as the cli version 0.10.8. Maybe you can add to the doc the specific version needed to install?

raykyri replied:
I had cut another release to get some of the features in the Canvas CLI, but the CLI should work now. Peering with the CLI *might* not work depending on your local setup... I think it's probably best to focus on the Commonwealth side rather than trying to QA peering, since strictly speaking that's in the next PR (I started adding the docs to this PR as a work in progress but they are not really complete.) 

I'm working on merging up to `master` which now includes the new API, which is holding up getting actions to propagate from the Canvas side to the Common side.

kurtisassad commented on packages/commonwealth/server/routes/threads/create_thread_comment_handler.ts:
```diff
@@ -58,43 +69,49 @@ export const createThreadCommentHandler = async (
+  // publish signed data
+  if (hasCanvasSignedDataApiArgs(req.body)) {
+    const { canvasSignedData } = fromCanvasSignedDataApiArgs(req.body);
+    await applyCanvasSignedData(canvas, canvasSignedData);
```
Since these are being written to a separate database, they are not able to be in the same transaction so we have a possibility of data inconsistency where the federation db is out of sync with the commonwealth db. I don't really understand the purpose of the federation db, but do we have some plan on how we will reconcile them in the future?

raykyri replied:
We need some kind of persistent data store (either Postgres or block storage/disk) to maintain a copy of the Merkle tree used for Canvas federation. Currently a separate Postgres db is the easiest way to get that.

It should probably be a separate database from Commonwealth, so that in case an external user pushes a bunch of data to it, that doesn't end up impacting Commonwealth uptime or or storage caps, and we have a clear boundary where we can implement rate limits.

kurtisassad reviewed (commented):
Can you add the deployment-plan tag to this PR and mention we need to add the federation env vars to heroku?

raykyri commented on libs/shared/src/canvas/runtime/node.ts:
```diff
@@ -0,0 +1,76 @@
+        data.sessionMessageSignature,
+        data.sessionMessage,
+      );
+      console.log('applied canvas session:', idSession);
```
Fixed - I had to move some of this code to packages/commonwealth/server/federation too (cleans things up and avoids a circular import)

raykyri commented on libs/shared/src/canvas/runtime/contract.ts:
```diff
@@ -0,0 +1,159 @@
+        updated_at: timestamp,
+      });
+    },
+    // TODO: not implemented (packages/commonwealth/server/routes/comments/update_comment_handler.ts)
```
Yeah, I want to get to this in a follow-up. Currently the edit thread/comment controllers only send select fields to the server, so I'd have to change around the Canvas contract to make it play well (and want to wait till after tRPC refactoring is done to try that).

raykyri commented on packages/commonwealth/client/scripts/state/api/comments/editComment.ts:
```diff
@@ -16,21 +16,22 @@ interface EditCommentProps {
   communityId,
-  parentCommentId,
-  threadId,
+  // parentCommentId,
```
Done

raykyri commented:
Okay, addressed comments, and merged up to master, and reconciled with the Create Thread tRPC route.

mzparacha commented:
Clicked around, Not able to create thread with active session address

https://github.com/user-attachments/assets/b90aaf43-978c-4f67-a2bb-0d5786c89c38

other actions with existing thread/comments/reactions worked for me.

raykyri commented:
I'll re-QA all the routes this week. 

For tests, changes in this PR need to be ported over to the new API. I've done a first pass of that (and will double check during QA), cc @rjwebb who is helping with fixing tests: https://github.com/hicommonwealth/commonwealth/pull/9110

raykyri commented:
(To clarify, the merge up to `master` is on #9110 until I finish a first round of QA)

raykyri commented:
QA matrix:

- [ ] Create and edit thread (from working session key)
- [ ] Create and edit comment (from working session key, on signed thread)
- [ ] Create and edit comment (from working session key, on unsigned thread)
- [ ] Create and delete reaction (from working session key, on signed thread)
- [ ] Create and delete reaction (from working session key, on unsigned thread)
- [ ] Create and delete reaction (from working session key, on signed comment)
- [ ] Create and delete reaction (from working session key, on unsigned comment)

Session re-auth test:

- [ ] Create thread (from cleared session key)
- [ ] Edit thread (from cleared session key)
- [ ] Create comment (from cleared session key)
- [ ] Edit comment (from cleared session key)
- [ ] Create reaction (on thread, from cleared session key)
- [ ] Delete reaction (on thread, from cleared session key)
- [ ] Create reaction (on comment, from cleared session key)
- [ ] Delete reaction (on comment, from cleared session key)

Multichain test:

- [ ] Create thread, comment, react thread, react comment, edit comment, edit thread, delete all (Solana)
- [ ] Create thread, comment, react thread, react comment, edit comment, edit thread, delete all (Evmos)
- [ ] Create thread, comment, react thread, react comment, edit comment, edit thread, delete all (Injective)
- [ ] Create thread, comment, react thread, react comment, edit comment, edit thread, delete all (Terra)
- [ ] Create thread, comment, react thread, react comment, edit comment, edit thread, delete all (Substrate, e.g. Edgeware)
- [ ] Create thread, comment, react thread, react comment, edit comment, edit thread, delete all (Cosmos, e.g. Stargaze)
- [ ] Create thread, comment, react thread, react comment, edit comment, edit thread, delete all (Ethereum, e.g. Aave)

mzparacha commented:
Facing the same issue as earlier https://github.com/hicommonwealth/commonwealth/pull/8537#issuecomment-2334208520

> Clicked around, Not able to create thread with active session address
 Screen.Recording.2024-09-06.at.7.27.39.PM.mov 
other actions with existing thread/comments/reactions worked for me.

Also got some auth issues with metamask/keplr (but they are also on master branch).

raykyri commented:
This should have fixed both the createThread and auth issues. If you're a super admin, you should disable that flag before testing this PR, or otherwise Roger has a PR that will be merged into master soon to fix a potential issue with super admin tRPC requests.

mzparacha commented:
Update:

Now all wallet logins work both inside and outside the community.

But I am still facing [this](https://github.com/hicommonwealth/commonwealth/pull/8537#issuecomment-2334208520) thread creation issue with all wallets.

Also the API build was failing for me (and per @raykyri it might be the cause of the create thread issue)
```
pnpm -r build
Scope: 15 of 16 workspace projects
libs/api-client build$ tsc -b ./tsconfig.build.json
└─ Done in 869ms
libs/chains build$ tsc -b tsconfig.build.json
│ src/cosmos-ts/src/codegen/helpers.ts(25,24): error TS2339: Property 'Buffer' does not exist on type 'typeof globalThis'.
│ src/cosmos-ts/src/codegen/helpers.ts(38,24): error TS2339: Property 'Buffer' does not exist on type 'typeof globalThis'.
└─ Failed in 2.7s at /commonwealth/libs/chains
```

will be attempting to fix that and update

mzparacha commented on packages/commonwealth/client/scripts/controllers/server/sessions.ts:
```diff
@@ -184,14 +198,15 @@ async function sign(
 }
 
 // Public signer methods
+// TODO: rename all these to `createSignThreadAction()` or similar
```
still a todo?

raykyri replied:
Yeah, trying not to make any more changes at this stage.

mzparacha commented:
1. Join community with address A -> Create thread -> `localStorage.clear()` -> try vote/comment/ -> login with address B that wasn't linked before -> notice error.

https://github.com/user-attachments/assets/1ec89623-d68b-4a28-9ba5-cb294cf71776
